File: app/src/main/java/com/tichuguru/ScoreHandFragment.java
package com.tichuguru;

import androidx.fragment.app.Fragment;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import com.tichuguru.model.Hand;
import com.tichuguru.model.Player;
import java.util.List;
import kankan.wheel.widget.OnWheelChangedListener;
import kankan.wheel.widget.WheelView;
import kankan.wheel.widget.adapters.ArrayWheelAdapter;

public class ScoreHandFragment extends Fragment {
    private Hand hand;
    private WheelView outFirst;
    private WheelView score1;
    private WheelView score2;
    private TextView total1;
    private TextView total2;

    public static ScoreHandFragment newInstance(Hand hand) {
        ScoreHandFragment f = new ScoreHandFragment();
        Bundle args = new Bundle();
        args.putSerializable("newHand", hand);
        f.setArguments(args);
        return f;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.scorehand, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        requireActivity().setTitle("Score Hand");

        hand = (Hand) requireArguments().getSerializable("newHand");
        total1 = view.findViewById(R.id.scoreHandTotal1);
        total2 = view.findViewById(R.id.scoreHandTotal2);

        OnWheelChangedListener changeListener = (wheel, oldValue, newValue) -> {
            if (wheel != outFirst) {
                WheelView other = (wheel == score1) ? score2 : score1;
                int val = Hand.CARD_SCORE_OPTIONS[newValue];
                int otherVal = Hand.otherCardScore(val);
                if (val != 0 || other.getCurrentItem() != Hand.cardScoreIndex(200)) {
                    other.setCurrentItem(Hand.cardScoreIndex(otherVal));
                }
            }
            updateHandScore();
        };

        score1 = view.findViewById(R.id.scoreHandScore1);
        score1.setViewAdapter(new ArrayWheelAdapter<>(requireContext(), Hand.CARD_SCORE_OPTIONS));
        score1.addChangingListener(changeListener);

        score2 = view.findViewById(R.id.scoreHandScore2);
        score2.setViewAdapter(new ArrayWheelAdapter<>(requireContext(), Hand.CARD_SCORE_OPTIONS));
        score2.addChangingListener(changeListener);

        List<Player> players = TGApp.getGame().getPlayers();
        String[] names = new String[4];
        for (int i = 0; i < 4; i++) names[i] = players.get(i).getName();

        outFirst = view.findViewById(R.id.scoreHandOutFirst);
        outFirst.setViewAdapter(new ArrayWheelAdapter<>(requireContext(), names));
        outFirst.addChangingListener(changeListener);

        view.findViewById(R.id.scoreHandSave).setOnClickListener(v -> onSave());

        ((TextView) view.findViewById(R.id.scoreHandName1)).setText(players.get(0).getName());
        ((TextView) view.findViewById(R.id.scoreHandName2)).setText(players.get(1).getName());
        ((TextView) view.findViewById(R.id.scoreHandName3)).setText(players.get(2).getName());
        ((TextView) view.findViewById(R.id.scoreHandName4)).setText(players.get(3).getName());

        score1.setCurrentItem(Hand.cardScoreIndex(50));
        score2.setCurrentItem(Hand.cardScoreIndex(50));
        for (int i = 0; i < 4; i++) {
            if (hand.isTichuFor(i) || hand.isGrandTichuFor(i)) {
                outFirst.setCurrentItem(i);
                break;
            }
        }
        updateHandScore();
    }

    private void updateHandScore() {
        hand.setCardScore1(Hand.CARD_SCORE_OPTIONS[score1.getCurrentItem()]);
        hand.setCardScore2(Hand.CARD_SCORE_OPTIONS[score2.getCurrentItem()]);
        hand.setOutFirst(outFirst.getCurrentItem());
        total1.setText(String.valueOf(hand.getTotalScore1()));
        total2.setText(String.valueOf(hand.getTotalScore2()));
    }

    private void onSave() {
        TGApp app = (TGApp) requireActivity().getApplication();
        TGApp.getGame().scoreHand(hand);
        app.saveGames();
        app.savePlayers();
        getParentFragmentManager().setFragmentResult("score_hand", new Bundle());
        getParentFragmentManager().popBackStack();
    }
}

File: app/src/main/kotlin/com/tichuguru/ScoreHandFragment.kt
package com.tichuguru

import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.NumberPicker
import android.widget.TextView
import androidx.core.os.BundleCompat
import com.tichuguru.model.Hand

class ScoreHandFragment : Fragment() {
    private lateinit var hand: Hand
    private lateinit var outFirst: NumberPicker
    private lateinit var score1: NumberPicker
    private lateinit var score2: NumberPicker
    private lateinit var total1: TextView
    private lateinit var total2: TextView

    companion object {
        private const val ARG_HAND         = "hand"
        private const val ARG_PLAYER_NAMES = "playerNames"

        fun newInstance(hand: Hand, playerNames: Array<String>): ScoreHandFragment {
            return ScoreHandFragment().apply {
                arguments = Bundle().apply {
                    putSerializable(ARG_HAND, hand)
                    putStringArray(ARG_PLAYER_NAMES, playerNames)
                }
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return inflater.inflate(R.layout.scorehand, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        requireActivity().title = "Score Hand"

        hand = BundleCompat.getSerializable(requireArguments(), ARG_HAND, Hand::class.java)!!
        val playerNames = requireArguments().getStringArray(ARG_PLAYER_NAMES)!!

        total1 = view.findViewById(R.id.scoreHandTotal1)
        total2 = view.findViewById(R.id.scoreHandTotal2)

        val scoreLabels = Array(Hand.CARD_SCORE_OPTIONS.size) { Hand.CARD_SCORE_OPTIONS[it].toString() }

        val changeListener = NumberPicker.OnValueChangeListener { picker, _, _ ->
            if (picker != outFirst) {
                val other = if (picker == score1) score2 else score1
                val v = Hand.CARD_SCORE_OPTIONS[picker.value]
                val otherVal = Hand.otherCardScore(v)
                if (v != 0 || other.value != Hand.cardScoreIndex(200)) {
                    other.value = Hand.cardScoreIndex(otherVal)
                }
            }
            updateHandScore()
        }

        score1 = view.findViewById(R.id.scoreHandScore1)
        score1.minValue = 0
        score1.maxValue = Hand.CARD_SCORE_OPTIONS.size - 1
        score1.displayedValues = scoreLabels
        score1.wrapSelectorWheel = false
        score1.setOnValueChangedListener(changeListener)

        score2 = view.findViewById(R.id.scoreHandScore2)
        score2.minValue = 0
        score2.maxValue = Hand.CARD_SCORE_OPTIONS.size - 1
        score2.displayedValues = scoreLabels
        score2.wrapSelectorWheel = false
        score2.setOnValueChangedListener(changeListener)

        outFirst = view.findViewById(R.id.scoreHandOutFirst)
        outFirst.minValue = 0
        outFirst.maxValue = 3
        outFirst.displayedValues = playerNames
        outFirst.wrapSelectorWheel = true
        outFirst.setOnValueChangedListener(changeListener)

        view.findViewById<View>(R.id.scoreHandSave).setOnClickListener { onSave() }

        view.findViewById<TextView>(R.id.scoreHandName1).text = playerNames[0]
        view.findViewById<TextView>(R.id.scoreHandName2).text = playerNames[1]
        view.findViewById<TextView>(R.id.scoreHandName3).text = playerNames[2]
        view.findViewById<TextView>(R.id.scoreHandName4).text = playerNames[3]

        score1.value = Hand.cardScoreIndex(50)
        score2.value = Hand.cardScoreIndex(50)
        for (i in 0..3) {
            if (hand.isTichuFor(i) || hand.isGrandTichuFor(i)) {
                outFirst.value = i
                break
            }
        }
        updateHandScore()
    }

    private fun updateHandScore() {
        hand.setCardScore1(Hand.CARD_SCORE_OPTIONS[score1.value])
        hand.setCardScore2(Hand.CARD_SCORE_OPTIONS[score2.value])
        hand.setOutFirst(outFirst.value)
        total1.text = hand.totalScore1.toString()
        total2.text = hand.totalScore2.toString()
    }

    private fun onSave() {
        ViewModelProvider(requireActivity())[TGViewModel::class.java].scoreHand(hand)
        parentFragmentManager.setFragmentResult("score_hand", Bundle())
        parentFragmentManager.popBackStack()
    }
}
